filestore.go

```diff
@@ -0,0 +1,282 @@
+	arrayURI string   // the uri of the array
+}
+
+// NewFile returns a file. The arrayURI is the uri of the underlying TileDB array that contains the file (see Export) or
+// in which the file will be stored (see Import)
+func NewFile(tdbCtx *Context, arrayURI string) *File {
+	return &File{
+		tdbCtx:   tdbCtx,
+		arrayURI: arrayURI,
+	}
+}
+
+// Size returns the uncompressed size of the file.
+func (f *File) Size() (int64, error) {
+	uri := C.CString(f.arrayURI)
+	defer C.free(unsafe.Pointer(uri))
+
+	var size C.size_t
+	ret := C.tiledb_filestore_size(f.tdbCtx.tiledbContext, uri, &size)
+	if ret != C.TILEDB_OK {
+		return 0, fmt.Errorf("Error getting file size: %s", f.tdbCtx.LastError())
+	}
+
+	return int64(size), nil
+}
+
+// Export writes the contents of f to the named local file.
+func (f *File) Export(name string) error {
+	arrayURI := C.CString(f.arrayURI)
+	defer C.free(unsafe.Pointer(arrayURI))
+	fileURI := C.CString(name)
+	defer C.free(unsafe.Pointer(fileURI))
+
+	ret := C.tiledb_filestore_uri_export(f.tdbCtx.tiledbContext, fileURI, arrayURI)
+	if ret != C.TILEDB_OK {
+		return fmt.Errorf("Error exporting file: %s", f.tdbCtx.LastError())
+	}
+
+	return nil
+}
+
+// Import replaces the contents of f with the contents of the named local file.
+// Assumes that the TileDB array exists and is of the appropriate schema.
+func (f *File) Import(name string, mimeType FileStoreMimeType) error {
+	arrayURI := C.CString(f.arrayURI)
+	defer C.free(unsafe.Pointer(arrayURI))
+	fileURI := C.CString(name)
+	defer C.free(unsafe.Pointer(fileURI))
+
+	ret := C.tiledb_filestore_uri_import(f.tdbCtx.tiledbContext, arrayURI, fileURI, C.tiledb_mime_type_t(mimeType))
+	if ret != C.TILEDB_OK {
+		return fmt.Errorf("Error importing file: %s", f.tdbCtx.LastError())
+	}
+
+	return nil
+}
+
+// CreateAndImport creates the array, initializes it with the appropriate schema and imports the file.
+func (f *File) CreateAndImport(name string, mimeType FileStoreMimeType) error {
+	schema, err := CreateFileSchemaFor(f.tdbCtx, name)
+	if err != nil {
+		return err
+	}
+	array, err := NewArray(f.tdbCtx, f.arrayURI)
+	if err != nil {
+		return err
+	}
+	err = array.Create(schema)
+	if err != nil {
+		return err
+	}
+
+	return f.Import(name, mimeType)
+}
+
+// FileReader is an io.Reader for filestore files.
+type FileReader struct {
+	tdbCtx    *Context // the tiledb context for all read operations
+	arrayURI  string   // the uri of the array
+	arraySize int64    // the size of the array as returned by GetSize
+	bytesRead int64    // the total bytes read so far. Used as an offset for read operations
+
+	cArrayURI *C.char // a malloced char* C string for arrayURI. Cached here to avoid malloc on every read.
+}
+
+// NewFileReader returns a FileReader for the array
+// ArrayURI is the uri of an existing array with the filestore schema.
+func NewFileReader(tdbCtx *Context, arrayURI string) (*FileReader, error) {
+	arraySize, err := NewFile(tdbCtx, arrayURI).Size()
```

Comment:
I think consensus was that this should not be calculated as part of the constructor because it could be an expensive operation and size isn't necessary until later. I think vfs.IsDir or IsFile or something along those lines are more prudent to check existance